### PR TITLE
Add optimizer priming for dist chkpt

### PR DESCRIPTION
**Pull request by jonb377**
See also: https://github.com/pytorch/xla/issues/6546

The optimizer state must be primed before it can be restored. Optimizer state isn't materialized until the first `optim.step` call, so to restore optimizer state before resuming training, a dummy step is needed.

This PR introduces the `prime_optimizer` API, which will run a dummy optimizer step with zeroed gradients. The gradient sharding is copied from the parameters to ensure the resulting sharding is the same.

### CODE REVIEW

**Review by alanwaketan — APPROVED** [review]
LGTM! Thanks, Jon.

**Comment by jonb377** [comment]
Hi @dasoto, I've found that this approach will not guarantee the same sharding in the optimizer compared to running a full training step (this is due to sharding propagation decisions in the compiler). I believe the AdamW unit test was broken after an openxla pin update, for example.

Since this is an experimental feature, I would be OK to merge after a rebase. cc @JackCaoG @alanwaketan
